### PR TITLE
Sending existing object to the webhook for the DELETE verb

### PR DESCRIPTION
**Pull request by caesarxuchao**
Based on #66535, with these major changes:
* The deleteValidation is plumbed to the GuaranteedUpdate and etcd3#Delete, so that it's always called when the update and delete retry on conflicts.

/assign
/sig api-machinery
/kind bug

```release-notes
For admission webhooks registered for DELETE operations on k8s built APIs or CRDs, the apiserver now sends the existing object as admissionRequest.Request.OldObject to the webhook. 
For custom apiservers they uses the generic registry in the apiserver library, they get this behavior automatically.
```

### CODE REVIEW

File: test/e2e/apimachinery/webhook.go [hunk 1]
```diff
@@ -198,13 +206,6 @@ var _ = SIGDescribe("AdmissionWebhook", func() {
 		testMultiVersionCustomResourceWebhook(f, testcrd)
 	})
 
-	It("Should deny crd creation", func() {
```
Comment (caesarxuchao) [inline test/e2e/apimachinery/webhook.go]: This is a duplicate of line 183.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go [hunk 2]
```diff
@@ -227,10 +199,34 @@ func (s *store) conditionalDelete(ctx context.Context, key string, out runtime.O
...
 		origState, err := s.getState(getResp, key, v, false)
-		if err != nil {
+		if err != nil && !storage.IsTransformerError(err) {
 			return err
 		}
-		if err := preconditions.Check(key, origState.obj); err != nil {
+		if err != nil && storage.IsTransformerError(err) && preconditions != nil {
+			return err
+		}
+		if err != nil && storage.IsTransformerError(err) && preconditions == nil {
```
Comment (caesarxuchao) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: This might be controversial.

File: test/e2e/apimachinery/webhook.go [hunk 3]
```diff
@@ -755,6 +756,36 @@ func testWebhook(f *framework.Framework) {
 	Expect(err).NotTo(HaveOccurred(), "failed to create configmap %s in namespace: %s", configmap.Name, skippedNamespaceName)
 }
 
+func testBlockingConfigmapDeletion(f *framework.Framework) {
+	By("create a configmap that should be denied by the webhook")
```
Comment (roycaihw) [inline test/e2e/apimachinery/webhook.go]: the text reads ambiguous to me. The creation should not be denied but the deletion should be

File: test/e2e/apimachinery/webhook.go [hunk 4]
```diff
@@ -1191,6 +1233,28 @@ func updateConfigMap(c clientset.Interface, ns, name string, update updateConfig
 	return cm, pollErr
 }
 
+type updateCustomResourceFn func(cm *unstructured.Unstructured)
+
+func updateCustomResource(c dynamic.ResourceInterface, ns, name string, update updateCustomResourceFn) (*unstructured.Unstructured, error) {
+	var cr *unstructured.Unstructured
+	pollErr := wait.PollImmediate(2*time.Second, 1*time.Minute, func() (bool, error) {
```
Comment (roycaihw) [inline test/e2e/apimachinery/webhook.go]: nit: I realized it's copied from existing method, but is 1 minute necessary?

File: pkg/registry/core/namespace/storage/storage.go [hunk 5]
```diff
@@ -175,6 +175,9 @@ func (r *REST) Delete(ctx context.Context, name string, options *metav1.DeleteOp
 					// wrong type
 					return nil, fmt.Errorf("expected *api.Namespace, got %v", existing)
 				}
+				if err := deleteValidation(existingNamespace); err != nil {
```
[the remark this answers is not in the document]
Comment (liggitt) [reply pkg/registry/core/namespace/storage/storage.go]: the old object passed to admission is the object currently persisted in etcd, prior to being affected by any of the mutations of the current request

File: pkg/registry/core/service/storage/rest.go [hunk 6]
```diff
@@ -233,7 +233,7 @@ func (rs *REST) Delete(ctx context.Context, id string, options *metav1.DeleteOpt
 	if !dryrun.IsDryRun(options.DryRun) {
 		// TODO: can leave dangling endpoints, and potentially return incorrect
 		// endpoints if a new service is created with the same name
-		_, _, err = rs.endpoints.Delete(ctx, id, &metav1.DeleteOptions{})
+		_, _, err = rs.endpoints.Delete(ctx, id, rest.ValidateAllObjectFunc, &metav1.DeleteOptions{})
```
Comment (liggitt) [inline pkg/registry/core/service/storage/rest.go]: interesting... does this mean delete admission is not (and was not) done on endpoints deleted by the service REST handler?

File: staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go [hunk 7]
```diff
@@ -863,16 +867,17 @@ func (e *Store) updateForGracefulDeletionAndFinalizers(ctx context.Context, name
...
-func (e *Store) Delete(ctx context.Context, name string, options *metav1.DeleteOptions) (runtime.Object, bool, error) {
+func (e *Store) Delete(ctx context.Context, name string, deleteValidation rest.ValidateObjectFunc, options *metav1.DeleteOptions) (runtime.Object, bool, error) {
 	key, err := e.KeyFunc(ctx, name)
 	if err != nil {
 		return nil, false, err
 	}
 	obj := e.NewFunc()
 	qualifiedResource := e.qualifiedResourceFromContext(ctx)
-	if err := e.Storage.Get(ctx, key, "", obj, false); err != nil {
+	if err = e.Storage.Get(ctx, key, "", obj, false); err != nil {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go]: was this change significant?

File: staging/src/k8s.io/apiserver/pkg/registry/rest/delete.go [hunk 8]
```diff
@@ -140,3 +141,39 @@ func BeforeDelete(strategy RESTDeleteStrategy, ctx context.Context, obj runtime.
 	}
 	return true, false, nil
 }
+
+// AdmissionToValidateObjectDeleteFunc returns a admission validate func for object deletion
+func AdmissionToValidateObjectDeleteFunc(admit admission.Interface, staticAttributes admission.Attributes, objInterfaces admission.ObjectInterfaces) ValidateObjectFunc {
+	mutatingAdmission, isMutatingAdmission := admit.(admission.MutationInterface)
+	validatingAdmission, isValidatingAdmission := admit.(admission.ValidationInterface)
+
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/registry/rest/delete.go]: could hoist the `mutatingAdmission.Handles(admission.Delete)` check outside and short-circuit the expensive attribute building if it doesn't handle delete

File: staging/src/k8s.io/apiserver/pkg/storage/errors/storage.go [hunk 9]
```diff
@@ -30,7 +30,7 @@ func InterpretListError(err error, qualifiedResource schema.GroupResource) error
 		return errors.NewNotFound(qualifiedResource, "")
 	case storage.IsUnreachable(err):
 		return errors.NewServerTimeout(qualifiedResource, "list", 2) // TODO: make configurable or handled at a higher level
-	case storage.IsInternalError(err):
+	case storage.IsTransformerError(err) || storage.IsInternalError(err):
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/errors/storage.go]: why were these changes necessary?

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go (hunk 2 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: this is interesting. wouldn't malformed data in etcd also prevent us from listing and finding out what the problematic key was? I'm trying to understand if this special handling actually enables us to resolve the issue, or if it's just a step in that direction.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go [hunk 10]
```diff
@@ -227,10 +199,34 @@ func (s *store) conditionalDelete(ctx context.Context, key string, out runtime.O
...
+			// in this rare case, we skip checking
+			// validateDeletion, to give user a way to delete the
+			// key via the apiserver in case of data corruption.
+			startTime := time.Now()
+			// We don't check the transaction response as we know
+			// the transformer cannot interpret it anyway.
+			_, txnErr := s.client.KV.Txn(ctx).If().Then(
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: should still check `clientv3.Compare(clientv3.ModRevision(key), "=", origState.rev)` to make sure we're deleting the same revision we couldn't interpret

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: don't we need txnResp to check `txnResp.Succeeded`, even if we don't decode?

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go [hunk 11]
```diff
@@ -227,10 +199,34 @@ func (s *store) conditionalDelete(ctx context.Context, key string, out runtime.O
...
+				clientv3.OpDelete(key),
+			).Commit()
+			metrics.RecordEtcdRequestLatency("delete", getTypeName(out), startTime)
+			if txnErr != nil {
+				return txnErr
+			}
+			return err
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: if we got a decode error, then deleted unconditionally, and that delete succeeded, I wouldn't expect an error to be returned

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go (hunk 2 above)
Comment (caesarxuchao) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: Good point. The error message send by the LIST handler does include the key: https://github.com/kubernetes/kubernetes/blob/0889c3e4e9f3f95232301de5432df0e21b3ee2c1/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go#L610

So this special handling enables us to resolve the issue.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/delete.go [hunk 12]
```diff
@@ -114,28 +114,12 @@ func DeleteResource(r rest.GracefulDeleter, allowsOptions bool, scope *RequestSc
 			return
 		}
 
-		trace.Step("About to check admission control")
-		if admit != nil && admit.Handles(admission.Delete) {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/delete.go]: before we merge this, we need a test that exercises delete and deletecollection on all the resources and subresources that support those verbs (including custom resources) and ensures admission is called.https://github.com/kubernetes/kubernetes/pull/76677 is starting a test like that to exercise webhook admission... can wait for that or do something simpler with non-webhook admission as part of this PR.

File: staging/src/k8s.io/apiserver/pkg/registry/rest/delete.go [hunk 13]
```diff
@@ -140,3 +141,39 @@ func BeforeDelete(strategy RESTDeleteStrategy, ctx context.Context, obj runtime.
...
+			staticAttributes.GetOperation(),
+			staticAttributes.IsDryRun(),
+			staticAttributes.GetUserInfo(),
+		)
+		if isMutatingAdmission && mutatingAdmission.Handles(finalAttributes.GetOperation()) {
+			if err := mutatingAdmission.Admit(finalAttributes, objInterfaces); err != nil {
+				return err
+			}
+		}
+		if isValidatingAdmission && validatingAdmission.Handles(finalAttributes.GetOperation()) {
```
[the remark this answers is not in the document]
Comment (caesarxuchao) [reply staging/src/k8s.io/apiserver/pkg/registry/rest/delete.go]: it's true now. But I want to keep it dynamic to be future proof.

File: test/e2e/apimachinery/webhook.go (hunk 4 above)
Comment (caesarxuchao) [reply test/e2e/apimachinery/webhook.go]: I don't remember why we made it 1 minute in the first place. I prefer to fix them in a dedicated PR if necessary.

File: pkg/registry/core/service/storage/rest.go (hunk 6 above)
Comment (caesarxuchao) [reply pkg/registry/core/service/storage/rest.go]: Right. I think it's ok. If needed, the service webhook can check the endpoints when validating the service deletion.

File: staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go (hunk 7 above)
Comment (caesarxuchao) [reply staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go]: No, I think it's my IDE needs update.

File: staging/src/k8s.io/apiserver/pkg/storage/errors/storage.go (hunk 9 above)
Comment (caesarxuchao) [reply staging/src/k8s.io/apiserver/pkg/storage/errors/storage.go]: These functions translates storage layer error to API error. This PR differentiates the TransformerError from other InternalError at the storage layer, but the error returned to clients is still the 500 internal error.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go (hunk 10 above)
Comment (caesarxuchao) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: Good point. See the last commit. Note that I had to modify `getState()` as well, because otherwise origState.rev would be nil.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go (hunk 11 above)
Comment (caesarxuchao) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: The original code behaved the same. If you feel strongly, I'd prefer fix it in another pull.

**Comment by liggitt** [comment]
KEP update and https://github.com/kubernetes/kubernetes/pull/76849 are merged, this should be good to rebase. You'll need to update the integration test "delete" tests to expect an old object now, but we now have the coverage we need to ensure no admission calls got dropped

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/etcd.go [hunk 14]
```diff
@@ -127,6 +127,9 @@ func (r *REST) Delete(ctx context.Context, name string, deleteValidation rest.Va
 				if existingCRD.DeletionTimestamp.IsZero() {
 					now := metav1.Now()
 					existingCRD.DeletionTimestamp = &now
+					// Setting the grace period to 0, so that once the finalizers are removed from the crd, apiserver deletes the crd.
+					var zero int64 = 0
+					existingCRD.DeletionGracePeriodSeconds = &zero
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/etcd.go]: this shouldn't be required, since CRD types don't implement graceful deletion, right?

File: test/integration/apiserver/admissionwebhook/admission_test.go [hunk 15]
```diff
@@ -501,6 +502,59 @@ func testResourceDelete(c *testContext) {
 		c.t.Error(err)
 		return
 	}
+
```
Comment (liggitt) [inline test/integration/apiserver/admissionwebhook/admission_test.go]: Call `c.admissionHolder.verifyResult(c.t)` and then set another expectation if you want to test multiple delete calls are made and passed to admission

File: test/integration/apiserver/admissionwebhook/admission_test.go [hunk 16]
```diff
@@ -501,6 +502,59 @@ func testResourceDelete(c *testContext) {
...
+	if err != nil {
+		c.t.Error(err)
+		return
+	}
+	c.admissionHolder.expect(c.gvr, gvk(c.resource.Group, c.resource.Version, c.resource.Kind), v1beta1.Delete, obj.GetName(), obj.GetNamespace(), false, true)
+	err = c.client.Resource(c.gvr).Namespace(obj.GetNamespace()).Delete(obj.GetName(), &metav1.DeleteOptions{GracePeriodSeconds: &zero, PropagationPolicy: &background})
+	if err != nil {
+		c.t.Error(err)
+		return
+	}
```
Comment (liggitt) [inline test/integration/apiserver/admissionwebhook/admission_test.go]: verify and set another expectation here

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/etcd.go (hunk 14 above)
Comment (caesarxuchao) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/etcd.go]: If the GracePeriod isn't set to 0, then after an UPDATE that removes all the finalizers, the apiserver will not automatically delete-on-update. See this line in `shouldDeleteDuringUpdate`:

https://github.com/kubernetes/kubernetes/blob/36e06bbb1fd3fc430ff4f8990684bbd53cf80e79/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L405

Reading this again, maybe we should fix `shouldDeleteDuringUpdate` instead, the condition should be 
```golang
func shouldDeleteDuringUpdate() bool {
  noGracePeriod := oldMeta.GetDeletionGracePeriodSeconds() == nil || *oldMeta.GetDeletionGracePeriodSeconds() == 0
  return len(newMeta.GetFinalizers()) == 0 && oldMeta.DeletionTimestamp() != nil && noGracePeriod
}
```

File: staging/src/k8s.io/apiserver/pkg/registry/rest/rest.go [hunk 17]
```diff
@@ -208,7 +211,7 @@ type UpdatedObjectInfo interface {
 // ValidateObjectFunc is a function to act on a given object. An error may be returned
 // if the hook cannot be completed. An ObjectFunc may NOT transform the provided
 // object.
-type ValidateObjectFunc func(obj runtime.Object) error
+type ValidateObjectFunc = storage.ValidateObjectFunc
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/registry/rest/rest.go]: I think we were trying to avoid this package dependency

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go [hunk 18]
```diff
@@ -227,12 +199,30 @@ func (s *store) conditionalDelete(ctx context.Context, key string, out runtime.O
...
+		if err != nil && !storage.IsTransformerError(err) {
 			return err
 		}
-		if err := preconditions.Check(key, origState.obj); err != nil {
+		if err != nil && storage.IsTransformerError(err) && preconditions != nil {
 			return err
 		}
+		// In this rare case, we skip checking validateDeletion to give
+		// user a way to delete the key via the apiserver in case of
+		// data corruption.
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: I want to think about this a little bit more

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go [hunk 19]
```diff
@@ -709,12 +705,12 @@ func (s *store) getState(getResp *clientv3.GetResponse, key string, v reflect.Va
 			return nil, err
 		}
 	} else {
+		state.rev = getResp.Kvs[0].ModRevision
+		state.meta.ResourceVersion = uint64(state.rev)
 		data, stale, err := s.transformer.TransformFromStorage(getResp.Kvs[0].Value, authenticatedDataString(key))
 		if err != nil {
-			return nil, storage.NewInternalError(err.Error())
+			return state, storage.NewTransformerError(err.Error())
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: where are we using the returned data even in the presence of an error?

**Comment by caesarxuchao** [comment]
Integration test failed with

```
TestWebhookV1beta1/cr.bar.com.v1.foos/delete
admission_test.go:555: foos.cr.bar.com "cr1foo" not found
```
Can't reproduce locally. I'll retest see if it's flaky.

**Comment by caesarxuchao** [comment]
It fails with another resource this time:
```
TestWebhookV1beta1/rbac.authorization.k8s.io.v1alpha1.roles/delete
admission_test.go:555: roles.rbac.authorization.k8s.io "role1" not found
```

**Comment by liggitt** [comment]
Hrmm… do subtests run in parallel?

**Comment by liggitt** [comment]
> admission_test.go:555

That's not a line where an error is logged, right? Is that the current revision?

**Comment by liggitt** [comment]
yeah, StartRealMasterOrDie doesn't appear to segment data in etcd

**Comment by liggitt** [comment]
looks like we have a bit of a mess with the organically grown test server startup methods

cmd/kube-apiserver/app/testing#StartTestServer[OrDie]
  * takes etcd storage config, typically used with segmented storage so multiple tests can coexist
test/integration/etcd#StartRealMasterOrDie
  * Originally designed just for the etcd test
  * it does not segment data in etcd (because the etcd path needed stable etcd key paths to check)
  * it wipes out etcd data when run
  * it locks to prevent concurrent use with other StartRealMasterOrDie tests
  * It does set up a CRD server (but not an aggregator)

We can make the webhook test coexist well by switching it to StartTestServer. Done in last commit of https://github.com/liggitt/kubernetes/commits/delete-admission-objects

**Comment by caesarxuchao** [comment]
Thanks @liggitt for looking into the test failure. I agree with your analysis, but I still don't get why `TestWebhookV1beta1`, which is using `StartRealMasterOrDie`, was flaky.
* It shouldn't race with other integration tests that use `StartRealMasterOrDie`, as they are protected by same etcd lock.
* It shouldn't be affected by other tests using `StartTestServer`, as those tests use a segmented etcd space.

Anyway, I cherry-picked your commit to see if it fixes the race.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go (hunk 19 above)
Comment (caesarxuchao) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: Did you mean "why"?

The state only contains `state.rev`, which comes from `getResp.Kvs[0].ModRevision`. The transformer error only indicates that the `getResp.Kvs[0].Value` is corrupted.

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: I meant where. Wanted to see how we were checking the value was usable/non-nil. Documentation should be explicit about what parts of the returned object are valid in the presence of an error

Comment (caesarxuchao) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: Right. Added documentation in commit https://github.com/kubernetes/kubernetes/pull/76346/commits/4a5699ad2cb1cea337d5b60b7251787ed65fa728.

The `state.rev` is used in https://github.com/kubernetes/kubernetes/pull/76346/files#diff-0f11327dd3daafdfe8e326e895616f66R236 in case of TranformError is returned.

**Comment by caesarxuchao** [comment]
Unfortunately @liggitt 's commit didn't fix the intergration failure. Test still failed with
```
TestWebhookV1beta1/apps.v1.replicasets/delete
admission_test.go:564: replicasets.apps "rs3" not found
```

I'll keep debugging.

**Comment by caesarxuchao** [comment]
Couldn't reproduce the issue after I added the logs..

**Comment by caesarxuchao** [comment]
Without the debugging logs, I reproduced the issue again:

```
admission_test.go:564: persistentvolumes "pv1name" not found
```

**Comment by caesarxuchao** [comment]
The most suspicious log from last failure was:

```
CHAO: finalizer for serviceaccounts/sa1name is [test/k8s.io]
CHAO: for serviceaccounts/sa1name, ignoreNotFound=false
 deleteImmediately=true
```

They are emitted by https://github.com/kubernetes/kubernetes/pull/76346/files#diff-2c081fad5c858e67c96f75adac185093R925 and https://github.com/kubernetes/kubernetes/pull/76346/files#diff-2c081fad5c858e67c96f75adac185093R941. It seems that another process removed the finalizer between the two lines.

**Comment by caesarxuchao** [comment]
The flake was [reproduced](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/76346/pull-kubernetes-integration/1125851182720356356/). Unfortunately, the failed resources were root scoped, and the additional logging I added was based on namespace, so it didn't give useful information.

**Comment by caesarxuchao** [comment]
Reproduced in this [run](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/76346/pull-kubernetes-integration/1125863388891582464/). Still don't know why the finalizer was removed. Adding more log to see if the object was updated.

**Comment by caesarxuchao** [comment]
So far what I found was that the finalizer was there at this [line](https://github.com/kubernetes/kubernetes/pull/76346/files#diff-2c081fad5c858e67c96f75adac185093R947), and somehow got removed before [line](https://github.com/kubernetes/kubernetes/pull/76346/files#diff-2c081fad5c858e67c96f75adac185093R951). I'm printing out the resource version to see if it's actually removed by another update.

**Comment by caesarxuchao** [comment]
```
CHAO: finalizer for roles.rbac.authorization.k8s.io/role1 is [test/k8s.io], uid is 01f1dcc1-65e4-4836-b85f-ccf0b3be4f56, rv is 1312
CHAO: again, finalizer for roles.rbac.authorization.k8s.io/role1 is [test/k8s.io], uid is 01f1dcc1-65e4-4836-b85f-ccf0b3be4f56, rv is 1312
CHAO: in updateForGracefulDeletionAndFinalizers 1, finalizer for role1 is [], uid is 01f1dcc1-65e4-4836-b85f-ccf0b3be4f56, rv is 1311
CHAO: in updateForGracefulDeletionAndFinalizers 2, finalizer for role1 is [], uid is 01f1dcc1-65e4-4836-b85f-ccf0b3be4f56, rv is 1311
CHAO: in updateForGracefulDeletionAndFinalizers 3, finalizer for role1 is [], uid is 01f1dcc1-65e4-4836-b85f-ccf0b3be4f56, rv is 1311
```

Somehow the rv goes backwards in this code block: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L873-L909.

**Comment by sttts** [comment]
Any chance to squash this into logical pieces for easier review?

File: test/integration/apiserver/admissionwebhook/admission_test.go [hunk 20]
```diff
@@ -99,6 +99,8 @@ var (
 	customTestFuncs = map[schema.GroupVersionResource]map[string]testFunc{
 		gvr("", "v1", "namespaces"): {"delete": testNamespaceDelete},
 
+		gvr("apiextensions.k8s.io", "v1beta1", "customresourcedefinitions"): {"delete": testCRDDelete},
```
Comment (caesarxuchao) [inline test/integration/apiserver/admissionwebhook/admission_test.go]: I had to add a specific delete test for CRD, because like namespaces, apiserver does not delete CRD on an update that removes the last pending finalizer.

Comment (liggitt) [reply test/integration/apiserver/admissionwebhook/admission_test.go]: should it? that seems like the same bug as the one you fixed in storage

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go [hunk 21]
```diff
@@ -741,10 +741,10 @@ func TestTransformationFailure(t *testing.T) {
 	}
 
 	// Delete succeeds but reports an error because we cannot access the body
-	if err := store.Delete(ctx, preset[1].key, &example.Pod{}, nil, storage.ValidateAllObjectFunc); !storage.IsInternalError(err) {
+	if err := store.Delete(ctx, preset[1].key, &example.Pod{}, nil, storage.ValidateAllObjectFunc); !storage.IsTransformerError(err) {
 		t.Errorf("Unexpected error: %v", err)
 	}
-	if err := store.Get(ctx, preset[1].key, "", &example.Pod{}, false); !storage.IsInternalError(err) {
+	if err := store.Get(ctx, preset[1].key, "", &example.Pod{}, false); !storage.IsNotFound(err) {
```
Comment (caesarxuchao) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go]: This matches the existing behavior: 

https://github.com/kubernetes/kubernetes/blob/08afeb855ffa9da78b862121aa38b0da7936dbe8/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go#L743-L750

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go [hunk 22]
```diff
@@ -227,12 +199,37 @@ func (s *store) conditionalDelete(ctx context.Context, key string, out runtime.O
...
-		if err := preconditions.Check(key, origState.obj); err != nil {
+		if err != nil && storage.IsTransformerError(err) && preconditions != nil {
+			// Checking preconditions requires origState.obj, which
+			// is not set in case of TransformerError.
+			// conditionalDelete fails closed in this case. If the
+			// user still wants to delete the object, they can
+			// retry the delete with empty preconditions.
 			return err
 		}
+		// We let conditionalDelete skip validateDeletion, delete the
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: Is it better to skip validateDeletion or call it with a nil object? Currently, delete admission is always called, regardless of transform error, with a nil oldObject. would it be preferable to continue calling delete admission with a nil oldObject in the case of a transform error over skipping it entirely?

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: I can't decide which bothers me more, giving an inconsistent API to admission consumers (*sometimes* getting a nil oldObject in a delete admission call), skipping delete admission entirely, or requiring talking to etcd to delete a malformed value.

Will CRD webhook conversion errors and KMS transform errors exercise this path? I'm not sure I'd want to abandon calling admission if an external system like that flakes.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go [hunk 23]
```diff
@@ -227,12 +199,37 @@ func (s *store) conditionalDelete(ctx context.Context, key string, out runtime.O
 	}
 	for {
 		origState, err := s.getState(getResp, key, v, false)
-		if err != nil {
+		if err != nil && !storage.IsTransformerError(err) {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: also return if `origState` is nil, since we don't have the rev info to use later

File: test/e2e/apimachinery/webhook.go (hunk 4 above)
Comment (liggitt) [reply test/e2e/apimachinery/webhook.go]: it tries every 2 seconds up to a minute... the minute timeout is expected to be generous

File: test/integration/apiserver/admissionwebhook/admission_test.go [hunk 24]
```diff
@@ -524,20 +528,76 @@ func testResourcePatch(c *testContext) {
...
 		c.t.Error(err)
 		return
 	}
+	c.admissionHolder.verify(c.t)
 
+	// remove the finalizer
+	_, err = c.client.Resource(c.gvr).Namespace(obj.GetNamespace()).Patch(
+		obj.GetName(),
+		types.MergePatchType,
+		[]byte(`{"metadata":{"finalizers":[]}}`),
```
Comment (liggitt) [inline test/integration/apiserver/admissionwebhook/admission_test.go]: should you wait until our finalizer is the only finalizer present to stomp remove all finalizers? this seems like it will race with the internal CRD finalizer

File: test/integration/apiserver/admissionwebhook/admission_test.go (hunk 20 above)
Comment (liggitt) [reply test/integration/apiserver/admissionwebhook/admission_test.go]: CRD storage doesn't override Update, and only changes Delete behavior on first delete... I don't think it should require a custom test method (I'd expect removing finalizers on update of a CRD with a deletiontimestamp set to actually delete)

Comment (caesarxuchao) [reply test/integration/apiserver/admissionwebhook/admission_test.go]: I removed the commit that modified `shouldDeleteDuringUpdate`. The function returns `false` unless the `DeletionGracePeriodSeconds` was set on the first Delete. CRD [storage](https://github.com/kubernetes/kubernetes/blob/4d3d153210c0788b8e188c467d31061470d8c042/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/etcd.go#L70) doesn't set the `DeletionGracePeriodSeconds`.

I removed my previous fix to `shouldDeleteDuringUpdate` because that would make  `Namespaces` misbehave.

I tracked the issue in https://github.com/kubernetes/kubernetes/issues/77944. We can solve it in a followup.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go (hunk 22 above)
Comment (caesarxuchao) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: Indeed the KMS transform is part of the transformer stack: https://github.com/kubernetes/kubernetes/blob/4d3d153210c0788b8e188c467d31061470d8c042/staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go#L190.

I'll trace the CRD webhook path later.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go (hunk 23 above)
Comment (caesarxuchao) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: Added a return after the error checks.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go (hunk 22 above)
Comment (caesarxuchao) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: Among all the options, "giving an inconsistent API to admission consumers" seems to be the lesser of the evils.

I added the last commit to send nil objects to `validateDelete`.

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: Discussed offline. since we already have to look up the existing object in the REST handler to verify there are no remaining finalizers, a persistent transform error is already blocking, and we'll never make it here. Continuing to simply return if we encounter a transform error (rather than skipping admission or passing a nil object) is no worse than the current state, and lets us give a consistent API to admission authors.

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: Let's drop the transformation error changes from this PR to keep it focused on passing existing objects to delete admission

File: test/integration/apiserver/admissionwebhook/admission_test.go (hunk 24 above)
Comment (caesarxuchao) [reply test/integration/apiserver/admissionwebhook/admission_test.go]: Ack. I added a wait loop. I don't know if the race with the internal CRD finalizer would cause any test flake, but we can't be too conservative.

**Comment by caesarxuchao** [comment]
@liggitt I cherry-picked your #77952 to get the integration tests working, I would need a rebase after #77952 merged. The PR is otherwise ready for another round of review. Thank you for the meticulous reviews.

**Comment by liggitt** [comment]
/lgtm
/approve

needs a rebase since https://github.com/kubernetes/kubernetes/pull/77952 merged

**Comment by yue9944882** [comment]
awesome! 🎉

**Comment by liggitt** [comment]
included in docs PR for 1.15 at https://github.com/kubernetes/website/pull/14671
